src/function-builder.ts

```diff
@@ -32,66 +32,99 @@ import * as https from './providers/https';
+ * List of all regions supported by Cloud Functions.
  */
-export function region(...regions: string[]) {
-  if (!regions.length) {
-    throw new Error('You must specify at least one region');
+const SUPPORTED_REGIONS = [
+  'us-central1',
+  'us-east1',
+  'europe-west1',
+  'europe-west2',
+  'asia-east2',
+  'asia-northeast1',
+];
+
+/**
+ * List of available memory options supported by Cloud Functions.
+ */
+const MEMORY_OPTS = ['128MB', '256MB', '512MB', '1GB', '2GB'];
+
+/**
+ * Cloud Functions max timeout value.
+ */
+const MAX_TIMEOUT_SECONDS = 540;
+
+/**
+ * Assert that the runtime options passed in are valid.
+ * @param runtimeOptions object containing memory and timeout information.
+ * @throws { Error } Memory and TimeoutSeconds values must be valid.
+ */
+function assertRuntimeOptionsValid(runtimeOptions: RuntimeOptions): boolean {
+  if (
+    runtimeOptions.memory &&
+    !_.includes(MEMORY_OPTS, runtimeOptions.memory)
+  ) {
+    throw new Error(
+      `The only valid memory allocation values are: ${MEMORY_OPTS.join(', ')}`
+    );
   }
   if (
-    _.difference(regions, [
-      'us-central1',
-      'us-east1',
-      'europe-west1',
-      'asia-northeast1',
-    ]).length
+    runtimeOptions.timeoutSeconds > MAX_TIMEOUT_SECONDS ||
+    runtimeOptions.timeoutSeconds < 0
   ) {
     throw new Error(
-      "The only valid regions are 'us-central1', 'us-east1', 'europe-west1', and 'asia-northeast1'"
+      `TimeoutSeconds must be between 0 and ${MAX_TIMEOUT_SECONDS}`
     );
   }
-  return new FunctionBuilder({ regions });
+  return true;
 }
+
+/**
+ * Assert regions specified are valid.
+ * @param regions list of regions.
+ * @throws { Error } Regions must be in list of supported regions.
+ */
+function assertRegionsAreValid(...regions: string[]): boolean {
+  if (!regions.length) {
+    throw new Error('You must specify at least one region');
+  }
+  if (_.difference(regions, SUPPORTED_REGIONS).length) {
+    throw new Error(
+      `The only valid regions are: ${SUPPORTED_REGIONS.join(', ')}`
+    );
+  }
+  return true;
+}
+
+/**
+ * Configure the regions that the function is deployed to.
+ * @param regions One of more region strings.
+ * For example: `functions.region('us-east1')` or `functions.region('us-east1', 'us-central1')`
+ */
+export function region(...regions: string[]): FunctionBuilder {
+  if (assertRegionsAreValid(...regions)) {
+    return new FunctionBuilder({ regions });
+  }
+}
+
 /**
  * Configure runtime options for the function.
  * @param runtimeOptions Object with 2 optional fields:
  * 1. `timeoutSeconds`: timeout for the function in seconds, possible values are 0 to 540
  * 2. `memory`: amount of memory to allocate to the function,
  *    possible values are:  '128MB', '256MB', '512MB', '1GB', and '2GB'.
  */
-export function runWith(runtimeOptions: {
-  timeoutSeconds?: number;
-  memory?: '128MB' | '256MB' | '512MB' | '1GB' | '2GB';
```

Comment:
should we keep this and maybe export a memory type as discussed?

Reply:
Done.